TACC/Core-CMS
# GH-75: Data List - Plugin

# To Do

- [x] [Use `get()` not direct indexing.](https://github.com/TACC/Core-CMS/pull/308/files/75b219c761ff5e14fd4062bc194b343a6a05c12f#r688134513)
- [x] (Optional) [Resolve list comprehension legibility.](https://github.com/TACC/Core-CMS/pull/308#discussion_r688064143)



# Overview

Implement "Data List" plugin.



# Issues

* #75



# Changes

- Create plugin for UI component.





# Screenshots

<details>
<summary>Form</summary>

| Plugin | Image |
| --- | --- |
| Data List | <img alt="GH-75 Form - Data List" src="https://user-images.githubusercontent.com/62723358/128584983-26cf5fd7-cb69-4945-b19e-611d2d0e5146.png" /> |
| Data List Item | <img width="799" alt="GH-75 Form - Data List Item" src="https://user-images.githubusercontent.com/62723358/129295203-2e113ebc-6b97-42be-ba54-b8730788cba3.png">[old screenshot](https://user-images.githubusercontent.com/62723358/128584989-375d1deb-34df-404d-b896-48203ec0d79d.png) |
| Link | either "Generic" > "Link" or "Bootstrap 4" > "Link / Button" |

</details>

<details>
<summary>Render</summary>

<img alt="GH-75 Render" src="https://user-images.githubusercontent.com/62723358/128584955-ff7ca5e4-045b-4d0e-bb2e-03ea2d35047d.jpeg" width="50%" />

</details>

<details>
<summary>Structure</summary>

<img alt="GH-75 Structure 1" src="https://user-images.githubusercontent.com/62723358/128584968-05c629c9-ad17-41fc-aaca-e1fbeb1254cc.png" width="50%" />
<img alt="GH-75 Structure 2" src="https://user-images.githubusercontent.com/62723358/128584975-6e4358b7-0fed-4e6d-9895-e8fbe8dfc726.png" width="50%" />

</details>





# Testing

> Last recorded deploy to [Frontera (dev)](https://dev.fronteraweb.tacc.utexas.edu/) was [Build #392 (Aug 9, 2021 1:02:37 PM)](https://jenkins01.tacc.utexas.edu/job/Core_Portal_Deploy/392/).

> Last recorded deploy to [Core (dev)](https://dev.cep.tacc.utexas.edu/) was [Build #393 (Aug 9, 2021 1:04:25 PM)](https://jenkins01.tacc.utexas.edu/job/Core_Portal_Deploy/393/).

1. If no environment, then [prepare to test CMS changes](https://github.com/TACC/Core-CMS/wiki/Test-CMS-Changes).
2. In CMS container, run `python3 manage.py migrate`.
3. On CMS website, logged in as admin, add "TACC Site" > "Data List" plugin.
3. Within the "Data List" plugin, add "TACC Site" > "Data List" plugin.
5. Ensure the following:
    1. each config value changes/adds a classname on the root element from the snippet markup ([classname/config map][constants])
    2. plugin description (next to name) in Structure view describes the required config values ([short_desc/config map][constants])
    3. labels and values added (via "Data List Item" plugins) are rendered
        - ✱ and an empty Label field results in no label text (an empty label spot)
    4. styles match those from https://github.com/TACC/Core-CMS/pull/305
        - ✱ but more space above & below between Table label/value cells
6. Within a "Data List Item":
    1. ✱ confirm "Link" fieldset instructions render well and make sense
    2. confirm "Advanced settings" > "Support child plugin for Label" is checked
    3. † add a "Generic" > "Link" ~~or "Bootstrap 4" > "Link / Button"~~‡ plugin
7. † Confirm label becomes the link you added.
8. Add "Generic" > "Text" plugin.
9. In Text plugin, in "CMS Plugins" dropdown, confirm these new plugins are **not** available:
    1. "TACC Site" > "Data List"
    2. "TACC Site" > "Data List Item"

✱ Step added since initial review.
† Step requires testing on a branch into which `task/GH-75-plugin` and `quick/support-plugins-unadulterated-by-bootstrap--for-main` have been merged.
‡ Removed since initial review, because this plugin is **not** supported in this PR.

[constants]: https://github.com/TACC/Core-CMS/blob/task/GH-75-plugin/taccsite_cms/contrib/taccsite_data_list/constants.py "Map of config values to description, short descritpion, and classname."
